core/src/rom_factory.cpp

```diff
@@ -0,0 +1,84 @@
+#include "core/rom_factory.h"
+
+#include "rom/nrom.h"
+
+#include <cstring>
+#include <fstream>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+static_assert(sizeof(n_e_s::core::INesHeader) == 16);
+
+namespace {
+
+std::istream::pos_type streamsize(std::istream &stream) {
+    stream.seekg(0, std::ios::end);
+    std::istream::pos_type size = stream.tellg();
+    stream.seekg(0, std::ios::beg);
+    return size;
+}
+
+} // namespace
+
+namespace n_e_s::core {
+
+std::unique_ptr<IRom> RomFactory::from_file(const std::string &filepath) {
+    std::ifstream file(filepath, std::ios::binary);
+    if (!file) {
+        throw std::invalid_argument("Unable to open file");
+    }
+
+    return from_bytes(file);
+}
+
+std::unique_ptr<IRom> RomFactory::from_bytes(std::istream &bytestream) {
+    std::vector<uint8_t> bytes(streamsize(bytestream));
+    if (bytes.size() < 16) {
+        throw std::invalid_argument(
+                "File isn't big enough to contain a header");
+    }
+
+    if (!bytestream.read(reinterpret_cast<char *>(&bytes[0]), bytes.size())) {
+        throw std::invalid_argument("Unable to get bytes");
+    }
+
+    INesHeader h;
+    if (!std::equal(bytes.begin(), bytes.begin() + sizeof(h.nes), h.nes)) {
+        throw std::invalid_argument("No valid iNes header");
+    }
+
+    // This is fine because the header is exactly 16 bytes with no padding.
+    memcpy(&h, &bytes[0], sizeof(h));
```

Comment:
The struct can contain padding though?

Reply:
No, the struct only contains unsigned chars which have no alignment requirement and thus will never cause the compiler to add any padding.

To make sure I didn't fuck that up, I also added a static_assert at the top of the file, asserting that the size of the struct is exactly 16 bytes which it'll only be if there's no padding. :P